lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,329 @@
+
+  for (uint32_t i = 0, e = structType.getNumElements(); i < e; ++i) {
+    auto memberType = SPIRVLayoutInfoUtils::decorateType(
+        structType.getElementType(i), structSize, memberAlignment);
+    memberTypes.push_back(memberType);
+    layoutInfo.push_back(memberOffset);
+    memberOffset = structSize;
+    // According to the SPIR-V spec:
+    // "A structure has a base alignment equal to the largest base alignment of
+    // any of its members."
+    maxMemberAlignment = std::max(maxMemberAlignment, memberAlignment);
+    memberAlignment = 1;
+  }
+
+  baseSize += llvm::alignTo(structSize, maxMemberAlignment);
+  baseAlignment = std::max(baseAlignment, maxMemberAlignment);
+
+  structType.getMemberDecorations(memberDecorations);
+  return spirv::StructType::get(memberTypes, layoutInfo, memberDecorations);
+}
+
+Type SPIRVLayoutInfoUtils::decorateType(
+    Type type, spirv::StructType::LayoutInfo &baseSize,
+    SPIRVLayoutInfoUtils::Alignment &baseAlignment) {
+
+  if (spirv::SPIRVDialect::isValidSPIRVScalarType(type)) {
+    auto scalarTypeAlignment =
+        SPIRVLayoutInfoUtils::getScalarTypeAlignment(type);
+    // SPIR-V spec does not specify any padding for a scalar type.
+    baseSize += scalarTypeAlignment;
+    baseAlignment = std::max(baseAlignment, scalarTypeAlignment);
+    return type;
+  }
+
+  switch (type.getKind()) {
+  case spirv::TypeKind::Struct:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<spirv::StructType>(),
+                                              baseSize, baseAlignment);
+  case spirv::TypeKind::Array:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<spirv::ArrayType>(),
+                                              baseSize, baseAlignment);
+  case StandardTypes::Vector:
+    return SPIRVLayoutInfoUtils::decorateType(type.cast<VectorType>(), baseSize,
+                                              baseAlignment);
+  default:
+    llvm_unreachable("unhandled SPIR-V type");
+  }
+}
+
+Type SPIRVLayoutInfoUtils::decorateType(
+    VectorType vectorType, spirv::StructType::LayoutInfo &baseSize,
+    SPIRVLayoutInfoUtils::Alignment &baseAlignment) {
+
+  const auto numElements = vectorType.getNumElements();
+  auto elementType = vectorType.getElementType();
+  spirv::StructType::LayoutInfo elementSize = 0;
+  SPIRVLayoutInfoUtils::Alignment elementAlignment = 1;
+
+  auto memberType = SPIRVLayoutInfoUtils::decorateType(elementType, elementSize,
+                                                       elementAlignment);
+  // According to the SPIR-V spec:
+  // 1."A two-component vector has a base alignment equal to twice its scalar
+  // alignment."
+  // 2."A three- or four-component vector has a base alignment equal to four
+  // times its scalar alignment."
+  auto vectorTypeAlignment =
+      numElements == 2 ? elementAlignment * 2 : elementAlignment * 4;
+
+  auto vectorSize = elementSize * numElements;
+  baseSize += llvm::alignTo(vectorSize, vectorTypeAlignment);
```

Comment:
So instead of letting each type to figure out how it should fit in the struct type by updating the `baseSize` and `baseAlignment` for the struct type, what about letting them focus on themselves? I mean for each of these functions, we still have the `size` and `alignment` parameters to "return" size and alignment, but make it just the size and alignment of the *current* type? It would avoid us to have all these `alignTo` and `max` calls in each case (we just need them in struct type). it's also easier to reason about alignment and padding if they are all in struct type, not scattered everywhere.

Reply:
@antiagainst as far as I understood, please fix me if I'm wrong, the spec allows to have nested structs,
something like this:
 ` !spv.ptr<!spv.struct<!spv.array<16x!spv.struct<f32, f64, !spv.array<16xf32>>>, f32>, StorageBuffer>`

in this case top level struct size must be aligned to 8 bytes, because:
>  An array has a base alignment equal to the base alignment of its element type.

> A structure has a base alignment equal to the largest base alignment of any of its members.

So, we have to go like DFS here, and update base alignment for the parent type to the largest alignment of its childs and in the same way replace types without layout to types with layout? Do I miss something?

Reply:
Yes the spec allows nested structs and we should recursively handle it. I meant that we don't need let child types to update for their parent type. Instead, they should just return their alignment and size and the parent type can do proper handling each child type. 

I think examples will make what I mean clear. You can check how layout is done in DXC at here: 

https://github.com/microsoft/DirectXShaderCompiler/blob/3e302950dd507e9c433705111ac0de1b143d6d1b/tools/clang/lib/SPIRV/AlignmentSizeCalculator.cpp#L66

Let me know if it's still not clear. Happy to clarify. :)

Reply:
@antiagainst thanks for the explanation, I hope I understood it right :)